source/fe/fe_nedelec_sz.cc

```diff
@@ -0,0 +1,1926 @@
+      //      the first line dof is 0,
+      //      the first face dof is n_line_dofs,
+      //      the first cell dof is n_line_dofs + n_face_dofs.
+      //
+      // On a line, DoFs are ordered first by line_dof and then line_index:
+      // i.e. line_dof_index = line_dof + line_index*(dofs_per_line)
+      //
+      // and similarly for faces:
+      // i.e. face_dof_index = face_dof + face_index*(dofs_per_face).
+      //
+      // HOWEVER, we have different types of DoFs on a line/face/cell.
+      // On a line we have two types, lowest order and higher order gradients.
+      //    - The numbering is such the lowest order is first, then higher order.
+      //      This is simple enough as there is only 1 lowest order and degree higher
+      //      orders DoFs per line.
+      //
+      // On a 2D cell, we have 3 types: Type 1/2/3:
+      //    - The ordering done by type:
+      //      - Type 1: 0 <= i1,j1 < degree. degree^2 in total.
+      //        Numbered: ij1 = i1 + j1*(degree).        i.e. cell_dof_index = ij1.
+      //      - Type 2: 0 <= i2,j2 < degree. degree^2 in total.
+      //        Numbered: ij2 = i2 + j2*(degree).        i.e. cell_dof_index = degree^2 + ij2
+      //      - Type 3: 0 <= i3 < 2*degree. 2*degree in total.
+      //        Numbered: ij3 = i3.                      i.e. cell_dof_index =  2*(degree^2) + ij3.
+      //
+      // These then fit into the local dof numbering described above:
+      // - local dof numberings are:
+      //   line_dofs: local_dof = line_dof_index.    0 <= local_dof < dofs_per_line*lines_per_cell
+      //   face_dofs: local_dof = n_line_dofs*lines_per_cell + face_dof_index.
+      //   cell dofs: local_dof = n_lines_dof + n_face_dofs + cell_dof_index.
+      //
+      // The cell-based shape functions are:
+      //
+      // Type 1 (gradients):
+      // \phi^{C_{1}}_{ij) = grad( L_{i+2}(2x-1)L_{j+2}(2y-1) ),
+      //
+      // 0 <= i,j < degree.
+      //
+      // NOTE: The derivative produced by IntegratedLegendrePolynomials does not account for the
+      //       (2*x-1) or (2*y-1) so we must take this into account when taking
+      //       derivatives.
+      const unsigned int cell_type1_offset = n_line_dofs;
+          
+      // Type 2:
+      // \phi^{C_{2}}_{ij) = L'_{i+2}(2x-1) L_{j+2}(2y-1) \mathbf{e}_{x}
+      //                     - L_{i+2}(2x-1) L'_{j+2}(2y-1) \mathbf{e}_{y},
+      //
+      // 0 <= i,j < degree.
+      const unsigned int cell_type2_offset = cell_type1_offset + degree*degree;
+          
+      // Type 3 (two subtypes):
+      // \phi^{C_{3}}_{j)        = L_{j+2}(2y-1) \mathbf{e}_{x}
+      //
+      // \phi^{C_{3}}_{j+degree) = L_{j+2}(2x-1) \mathbf{e}_{y}
+      //
+      // 0 <= j < degree
+      const unsigned int cell_type3_offset1 = cell_type2_offset + degree*degree;
+      const unsigned int cell_type3_offset2 = cell_type3_offset1 + degree;
+      
+      if (flags & update_values | update_gradients)
```

Comment:
Hm... Does this work as intented?
Maybe write:

`if (flags & (update_values | update_gradients))`